File: sample-objc/ARGEARSample/ARGearHelpers/BeautyManager.m
//
//  BeautyManager.m
//  ARGEARSample
//
//  Created by Jihye on 24/10/2019.
//  Copyright © 2019 Seerslab. All rights reserved.
//

#import "BeautyManager.h"

@interface BeautyManager() {
    float _beautyValue[ARGContentItemBeautyNum];
}

@end

@implementation BeautyManager

+ (BeautyManager *)shared {
    static dispatch_once_t pred;
    static BeautyManager *sharedInstance = nil;
    dispatch_once(&pred, ^{
        sharedInstance = [[BeautyManager alloc] init];
    });
    return sharedInstance;
}

- (instancetype)init {
    self = [super init];
    if (self) {
        
        _beautyRange200Array = @[
            @(ARGContentItemBeautyChin),
            @(ARGContentItemBeautyEyeGap),
            @(ARGContentItemBeautyNoseLength),
            @(ARGContentItemBeautyMouthSize),
            @(ARGContentItemBeautyEyeCorner),
            @(ARGContentItemBeautyLipSize)
        ];

        [self loadBeautyValue];
    }
    return self;
}


- (void)start {
    [[_session contents] setBulge:ARGContentItemBulgeNONE];
    [[_session contents] setDefaultBeauty];
}

- (void)setDefault {
    [[_session contents] setDefaultBeauty];
}

- (void)off {
    [self loadBeautyValue];
    
    [[_session contents] setBeautyOn:NO];
}

- (void)on {
    [[_session contents] setBeautyValues:_beautyValue];
}


- (void)setBeauty:(ARGContentItemBeauty)type value:(CGFloat)value {
    value = [self convertSliderValueToBeautyValue:type value:value];
    
    [[_session contents] setBeauty:type value:value];
}

- (CGFloat)getBeautyValue:(ARGContentItemBeauty)type {
    CGFloat value = [[_session contents] getBeautyValue:type];
    
    return [self convertBeautyValueToSliderValue:type value:value];
}

- (void)loadBeautyValue {
    for (NSInteger i = 0; i < ARGContentItemBeautyNum; i++) {
        _beautyValue[i] = [[_session contents] getBeautyValue:i];
    }
}

// 0 ~ 1 -> 0 ~ 100 or -100 ~ 100
- (CGFloat)convertSliderValueToBeautyValue:(ARGContentItemBeauty)type value:(CGFloat)value {
    
    if (value < 0) {
        value = 0.0f;
    }
    
    if (value > 1) {
        value = 1.0f;
    }
    
    if ([_beautyRange200Array containsObject:@(type)]) {
        value = (value * 200.f) - 100.f;
    } else {
        value = value * 100.f;
    }
    
    return value;
}

// 0 ~ 100 or -100 ~ 100 -> 0 ~ 1
- (CGFloat)convertBeautyValueToSliderValue:(ARGContentItemBeauty)type value:(CGFloat)value {
    
    if ([_beautyRange200Array containsObject:@(type)]) {
        value = (value + 100.f) / 200.f;
    } else {
        value = value / 100.f;
    }
    
    return value;
}

@end

File: sample-swift/ARGearSample/ARGearHelpers/BeautyManager.swift
//
//  BeautyManager.swift
//  ARGearSample
//
//  Created by Jihye on 2020/02/10.
//  Copyright © 2020 Seerslab. All rights reserved.
//

import Foundation
import ARGear

class BeautyManager {
    static let shared = BeautyManager()
    
    private var beautyValues: [Float] = Array(repeating: 0, count: ARGContentItemBeauty.num.rawValue)
    
    var argSession: ARGSession?
    var beautyRange200Array: [ARGContentItemBeauty] = [
        .chin,
        .eyeGap,
        .noseLength,
        .mouthSize,
        .eyeCorner,
        .lipSize
    ]
    
    init() {
        self.loadBeautyValue()
    }
    
    public func start() {
        guard let session = self.argSession, let contents = session.contents
            else { return }
        
        contents.setBulge(.NONE)
        contents.setDefaultBeauty()
    }
    
    public func setDefault() {
        guard let session = self.argSession, let contents = session.contents
            else { return }
        
        contents.setDefaultBeauty()
    }
    
    public func off() {
        guard let session = self.argSession, let contents = session.contents
            else { return }
        
        self.loadBeautyValue()
        
        contents.setBeautyOn(false)
    }
    
    public func on() {
        guard let session = self.argSession, let contents = session.contents
            else { return }
        
        let beautyValuePointer: UnsafeMutablePointer<Float> = UnsafeMutablePointer(&self.beautyValues)
        
        contents.setBeautyValues(beautyValuePointer)
    }
    
    public func setBeauty(type: ARGContentItemBeauty, value: Float) {
        guard let session = self.argSession, let contents = session.contents
            else { return }
        
        let value = self.convertSliderValueToBeautyValue(type: type, value: value)
        
        contents.setBeauty(type, value: value)
    }
    
    public func getBeautyValue(type: ARGContentItemBeauty) -> Float {
        guard let session = self.argSession, let contents = session.contents
            else { return 0 }
        
        let value = contents.getBeautyValue(type)
        
        return self.convertBeautyValueToSliderValue(type: type, value: value)
    }
    
    // load current ARGear's beauty values
    private func loadBeautyValue() {
        guard let session = self.argSession, let contents = session.contents
            else { return }

        for i in 0..<ARGContentItemBeauty.num.rawValue {
            self.beautyValues[i] = contents.getBeautyValue(ARGContentItemBeauty(rawValue: i)!)
        }
    }
    
    // 0 ~ 1 -> 0 ~ 100 or -100 ~ 100
    private func convertSliderValueToBeautyValue(type: ARGContentItemBeauty, value: Float) -> Float {
        var beautyValue = value
        if value < 0 {
            beautyValue = 0.0;
        }
        
        if value > 1 {
            beautyValue = 1.0;
        }
        
        if beautyRange200Array.contains(type) {
            beautyValue = (value * 200.0) - 100.0
        } else {
            beautyValue = value * 100.0
        }
        
        return beautyValue
    }
    
    // 0 ~ 100 or -100 ~ 100 -> 0 ~ 1
    private func convertBeautyValueToSliderValue(type: ARGContentItemBeauty, value: Float) -> Float {
        var beautyValue = value
        if beautyRange200Array.contains(type) {
            beautyValue = (value + 100.0) / 200.0
        } else {
            beautyValue = value / 100.0
        }
        
        return beautyValue
    }
}
